velox/common/memory/SharedArbitrator.h

```diff
@@ -85,92 +87,182 @@ class SharedArbitrator : public memory::MemoryArbitrator {
       "globalArbitrationCount"};
   static inline const std::string kLocalArbitrationCount{
       "localArbitrationCount"};
+  static inline const std::string kLocalArbitrationQueueWallNanos{
+      "localArbitrationQueueWallNanos"};
+  static inline const std::string kLocalArbitrationLockWaitWallNanos{
+      "localArbitrationLockWaitWallNanos"};
+  static inline const std::string kGlobalArbitrationLockWaitWallNanos{
+      "globalArbitrationLockWaitWallNanos"};
 
  private:
   // The kind string of shared arbitrator.
   inline static const std::string kind_{"SHARED"};
 
+  // Contains the execution state of an arbitration operation.
+  struct ArbitrationOperation {
+    MemoryPool* const requestPool;
+    MemoryPool* const requestRoot;
+    const std::vector<std::shared_ptr<MemoryPool>>& candidatePools;
+    const uint64_t targetBytes;
+    // The start time of this arbitration operation.
+    const std::chrono::steady_clock::time_point startTime;
+
+    // The stats of candidate memory pools used for memory arbitration.
+    std::vector<Candidate> candidates;
+    // The time that waits in local arbitration queue.
+    uint64_t localArbitrationQueueTimeUs{0};
+    // The time that waits to acquire the local arbitration lock.
+    uint64_t localArbitrationLockWaitTimeUs{0};
+    // The time that waits to acquire the global arbitration lock.
+    uint64_t globalArbitrationLockWaitTimeUs{0};
+
+    ArbitrationOperation(
+        uint64_t targetBytes,
+        const std::vector<std::shared_ptr<MemoryPool>>& candidatePools)
+        : ArbitrationOperation(nullptr, targetBytes, candidatePools) {}
+
+    ArbitrationOperation(
+        MemoryPool* _requestor,
+        uint64_t _targetBytes,
+        const std::vector<std::shared_ptr<MemoryPool>>& _candidatePools)
+        : requestPool(_requestor),
+          requestRoot(_requestor == nullptr ? nullptr : _requestor->root()),
+          candidatePools(_candidatePools),
+          targetBytes(_targetBytes),
+          startTime(std::chrono::steady_clock::now()) {}
+
+    uint64_t waitTimeUs() const {
+      return localArbitrationQueueTimeUs + localArbitrationLockWaitTimeUs +
+          globalArbitrationLockWaitTimeUs;
+    }
+
+    void enterArbitration();
+    void leaveArbitration();
+  };
+
+  // Used to start and finish an arbitration operation initiated from a memory
+  // pool or memory capacity shrink request sent through shrinkPools() API.
   class ScopedArbitration {
```

Comment:
If we decide not to move queueing out, can we make the comments more clear:
Prepare pool to enter arbitration state... put it to local arbitration queue... 
Exit pool from arbitration state... move local queue forward...

But I think we should consider bring the global queue back and move queue out from thie ScopedArbitration. See the other comments